src/mod_block_incoming.erl

```diff
@@ -0,0 +1,298 @@
+  end
+;
+
+handle_call({remove_blocking_user, User}, _From, {Host, BlockingUsers}) ->
+  {reply, ok, {Host, lists:delete(User, BlockingUsers)}}
+;
+
+handle_call({reset_blocking_users}, _From, {Host, _BlockingUsers}) ->
+  {reply, ok, { Host, gen_mod:get_module_opt(Host, ?MODULE, blocking_users, []) }}
+;
+
+handle_call({clear_blocking_users}, _From, {Host, _BlockingUsers}) ->
+  {reply, ok, { Host, [] }}
+;
+
+handle_call(Request, From, State) ->
+  ?ERROR_MSG("Unexpected handle_call from [~p] for [~p]", [From, Request]),
+  {noreply, State}
+.
+
+handle_cast(Request, State) ->
+  ?ERROR_MSG("Unexpected handle_cast: [~p]", [Request]),
+  {noreply, State}
+.
+
+handle_info(Info, State) ->
+  ?ERROR_MSG("Unexpected handle_info: [~p]", [Info]),
+  {noreply, State}
+.
+
+code_change(_OldVsn, { Host, _BlockingUsers } = State, _Extra) ->
+  ?DEBUG("Updating code for mod_block_incoming process on host [~p]", [Host]),
+  {ok, State}
+.
+
+%% filter messages
+filter_packet({#message{from = From} = Msg, State} = Acc) ->
+  LFrom = jid:tolower(From),
+  LBareFrom = jid:remove_resource(LFrom),
+  #{pres_a := PresA} = State,
+  FromUserInPresencePacket = gb_sets:is_element(LFrom, PresA) orelse gb_sets:is_element(LBareFrom, PresA) orelse sets_bare_member(LBareFrom, PresA),
+  case FromUserInPresencePacket of
+    false -> %% from-user not in presence struct so this is an outgoing DM to someone else
+      case check_message(Msg) of
+        allow -> Acc;
+        deny -> {stop, {drop, State}}
+      end;
+    true -> %% from-user is in the presence struct so either a self-message or a room chat with them in it
+      Acc
+  end
+;
+
+filter_packet(Acc) -> Acc.
+
+filter_offline_msg({_Action, #message{} = Msg} = Acc) ->
+  case check_message(Msg) of
+    allow -> Acc;
+    deny -> {stop, {drop, Msg}}
+  end
+.
+
+filter_subscription(Acc, #presence{from = From, to = To, type = subscribe} = Pres) ->
+  case need_check(Pres) of
+    true ->
+      case check_subscription(From, To) of
+        false -> {stop, false};
+        true -> Acc
+      end;
+    false -> Acc
+  end
+;
+
+filter_subscription(Acc, _) -> Acc.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+check_message(#message{from = From, to = To, lang = Lang} = Msg) ->
+  case need_check(Msg) of
+    true ->
+      case check_subscription(From, To) of
+        false ->
+          Txt = <<"Messages from strangers are rejected for this recipient">>,
+          Err = xmpp:err_policy_violation(Txt, Lang),
+          Msg1 = maybe_adjust_from(Msg),
+          ejabberd_router:route_error(Msg1, Err),
+          deny;
+        true -> allow
+      end;
+    false -> allow
+  end
+.
+
+check_subscription(From, To) ->
+  mod_roster:is_subscribed(From, To)
+.
+
+need_check(Pkt) ->
+  To = xmpp:get_to(Pkt),
+  From = xmpp:get_from(Pkt),
+
+  %% IsToAnotherUser:
```

Comment:
it might be more clear to declare these as variables and then string them together instead of commenting